`block-scoped-var` fails on deeper desctructuring

```javascript
let {PropTypes, addons: {PureRenderMixin}} = React;
let Test = React.createClass({
  mixins: [PureRenderMixin],

  // ...
});
```

```
  11:11  error  "PureRenderMixin" used outside of binding context  block-scoped-var
```

Moved out from closed issue https://github.com/eslint/eslint/issues/1863#issuecomment-89869081

I've updated to 0.19, but issue is still there.

### Output
```
21:11 $ node_modules/.bin/eslint src/test.js

src/test.js
  7:11  error  "PureRenderMixin" used outside of binding context  block-scoped-var

✖ 1 problem (1 error, 0 warnings)
```

### src/test.js

```javascript
'use strict';

const React = require('react/addons');
const {addons: {PureRenderMixin}} = React;

const Data = React.createClass({
  mixins: [PureRenderMixin],
  render() {
    return <div>Hello, world!</div>;
  }
});

module.exports = Data;

```


### .eslintrc

```javascript
{
  "env": {
    "browser": true,
    "node": true,
    "es6": true
  },

  "plugins": ["react"],

  "ecmaFeatures": {
    "arrowFunctions": true,
    "binaryLiterals": true,
    "blockBindings": true,
    "classes": true,
    "defaultParams": true,
    "destructuring": true,
    "forOf": true,
    "generators": true,
    "modules": true,
    "objectLiteralComputedProperties": true,
    "objectLiteralDuplicateProperties": true,
    "objectLiteralShorthandMethods": true,
    "objectLiteralShorthandProperties": true,
    "octalLiterals": true,
    "regexUFlag": true,
    "regexYFlag": true,
    "spread": true,
    "superInFunctions": true,
    "templateStrings": true,
    "unicodeCodePointEscapes": true,
    "globalReturn": true,
    "jsx": true
  },

  "rules": {

    //
    //Possible Errors
    //
    // The following rules point out areas where you might have made mistakes.
    //
    "comma-dangle": 2, // disallow or enforce trailing commas
    "no-cond-assign": 2, // disallow assignment in conditional expressions
    "no-console": 1, // disallow use of console (off by default in the node environment)
    "no-constant-condition": 2, // disallow use of constant expressions in conditions
    "no-control-regex": 2, // disallow control characters in regular expressions
    "no-debugger": 2, // disallow use of debugger
    "no-dupe-args": 2, // disallow duplicate arguments in functions
    "no-dupe-keys": 2, // disallow duplicate keys when creating object literals
    "no-duplicate-case": 2, // disallow a duplicate case label.
    "no-empty": 2, // disallow empty statements
    "no-empty-class": 2, // disallow the use of empty character classes in regular expressions
    "no-ex-assign": 2, // disallow assigning to the exception in a catch block
    "no-extra-boolean-cast": 2, // disallow double-negation boolean casts in a boolean context
    "no-extra-parens": 0, // disallow unnecessary parentheses (off by default)
    "no-extra-semi": 2, // disallow unnecessary semicolons
    "no-func-assign": 2, // disallow overwriting functions written as function declarations
    "no-inner-declarations": 2, // disallow function or variable declarations in nested blocks
    "no-invalid-regexp": 2, // disallow invalid regular expression strings in the RegExp constructor
    "no-irregular-whitespace": 2, // disallow irregular whitespace outside of strings and comments
    "no-negated-in-lhs": 2, // disallow negation of the left operand of an in expression
    "no-obj-calls": 2, // disallow the use of object properties of the global object (Math and JSON) as functions
    "no-regex-spaces": 2, // disallow multiple spaces in a regular expression literal
    "no-reserved-keys": 2, // disallow reserved words being used as object literal keys (off by default)
    "no-sparse-arrays": 2, // disallow sparse arrays
    "no-unreachable": 2, // disallow unreachable statements after a return, throw, continue, or break statement
    "use-isnan": 2, // disallow comparisons with the value NaN
    "valid-jsdoc": 2, // Ensure JSDoc comments are valid (off by default)
    "valid-typeof": 2, // Ensure that the results of typeof are compared against a valid string

    //
    // Best Practices
    //
    // These are rules designed to prevent you from making mistakes.
    // They either prescribe a better way of doing something or help you avoid footguns.
    //
    "block-scoped-var": 2, // treat var statements as if they were block scoped (off by default). 0: deep destructuring is not compatible https://github.com/eslint/eslint/issues/1863
    "complexity": 0, // specify the maximum cyclomatic complexity allowed in a program (off by default)
    "consistent-return": 2, // require return statements to either always or never specify values
    "curly": 2, // specify curly brace conventions for all control statements
    "default-case": 2, // require default case in switch statements (off by default)
    "dot-notation": 2, // encourages use of dot notation whenever possible
    "eqeqeq": 2, // require the use of === and !==
    "guard-for-in": 2, // make sure for-in loops have an if statement (off by default)
    "no-alert": 2, // disallow the use of alert, confirm, and prompt
    "no-caller": 2, // disallow use of arguments.caller or arguments.callee
    "no-div-regex": 2, // disallow division operators explicitly at beginning of regular expression (off by default)
    "no-else-return": 2, // disallow else after a return in an if (off by default)
    "no-empty-label": 2, // disallow use of labels for anything other then loops and switches
    "no-eq-null": 2, // disallow comparisons to null without a type-checking operator (off by default)
    "no-eval": 2, // disallow use of eval()
    "no-extend-native": 2, // disallow adding to native types
    "no-extra-bind": 2, // disallow unnecessary function binding
    "no-fallthrough": 2, // disallow fallthrough of case statements
    "no-floating-decimal": 2, // disallow the use of leading or trailing decimal points in numeric literals (off by default)
    "no-implied-eval": 2, // disallow use of eval()-like methods
    "no-iterator": 2, // disallow usage of __iterator__ property
    "no-labels": 2, // disallow use of labeled statements
    "no-lone-blocks": 2, // disallow unnecessary nested blocks
    "no-loop-func": 2, // disallow creation of functions within loops
    "no-multi-spaces": 2, // disallow use of multiple spaces
    "no-multi-str": 2, // disallow use of multiline strings
    "no-native-reassign": 2, // disallow reassignments of native objects
    "no-new": 2, // disallow use of new operator when not part of the assignment or comparison
    "no-new-func": 2, // disallow use of new operator for Function object
    "no-new-wrappers": 2, // disallows creating new instances of String,Number, and Boolean
    "no-octal": 2, // disallow use of octal literals
    "no-octal-escape": 2, // disallow use of octal escape sequences in string literals, such as var foo = "Copyright \251";
    "no-param-reassign": 2, // disallow reassignment of function parameters (off by default)
    "no-process-env": 2, // disallow use of process.env (off by default)
    "no-proto": 2, // disallow usage of __proto__ property
    "no-redeclare": 2, // disallow declaring the same variable more then once
    "no-return-assign": 2, // disallow use of assignment in return statement
    "no-script-url": 2, // disallow use of javascript: urls.
    "no-self-compare": 2, // disallow comparisons where both sides are exactly the same (off by default)
    "no-sequences": 2, // disallow use of comma operator
    "no-throw-literal": 2, // restrict what can be thrown as an exception (off by default)
    "no-unused-expressions": 2, // disallow usage of expressions in statement position
    "no-void": 2, // disallow use of void operator (off by default)
    "no-warning-comments": [0, {"terms": ["todo", "fixme"], "location": "start"}], // disallow usage of configurable warning terms in comments": 2, // e.g. TODO or FIXME (off by default)
    "no-with": 2, // disallow use of the with statement
    "radix": 2, // require use of the second argument for parseInt() (off by default)
    "vars-on-top": 2, // requires to declare all vars on top of their containing scope (off by default)
    "wrap-iife": 2, // require immediate function invocation to be wrapped in parentheses (off by default)
    "yoda": 2, // require or disallow Yoda conditions

    //
    // Strict Mode
    //
    // These rules relate to using strict mode.
    //
    "strict": 0, // controls location of Use Strict Directives. 0: required by `babel-eslint`

    //
    // Variables
    //
    // These rules have to do with variable declarations.
    //
    "no-catch-shadow": 2, // disallow the catch clause parameter name being the same as a variable in the outer scope (off by default in the node environment)
    "no-delete-var": 2, // disallow deletion of variables
    "no-label-var": 2, // disallow labels that share a name with a variable
    "no-shadow": 2, // disallow declaration of variables already declared in the outer scope
    "no-shadow-restricted-names": 2, // disallow shadowing of names such as arguments
    "no-undef": 2, // disallow use of undeclared variables unless mentioned in a /*global */ block
    "no-undef-init": 2, // disallow use of undefined when initializing variables
    "no-undefined": 2, // disallow use of undefined variable (off by default)
    "no-unused-vars": 2, // disallow declaration of variables that are not used in the code
    "no-use-before-define": 2, // disallow use of variables before they are defined

    //
    //Stylistic Issues
    //
    // These rules are purely matters of style and are quite subjective.
    //
    "indent": [1, 2], // this option sets a specific tab width for your code (off by default)
    "brace-style": 1, // enforce one true brace style (off by default)
    "camelcase": 1, // require camel case names
    "comma-spacing": [1, {"before": false, "after": true}], // enforce spacing before and after comma
    "comma-style": [1, "last"], // enforce one true comma style (off by default)
    "consistent-this": [1, "_this"], // enforces consistent naming when capturing the current execution context (off by default)
    "eol-last": 1, // enforce newline at the end of file, with no multiple empty lines
    "func-names": 0, // require function expressions to have a name (off by default)
    "func-style": 0, // enforces use of function declarations or expressions (off by default)
    "key-spacing": [1, {"beforeColon": false, "afterColon": true}], // enforces spacing between keys and values in object literal properties
    "max-nested-callbacks": [1, 3], // specify the maximum depth callbacks can be nested (off by default)
    "new-cap": [1, {"newIsCap": true, "capIsNew": false}], // require a capital letter for constructors
    "new-parens": 1, // disallow the omission of parentheses when invoking a constructor with no arguments
    "newline-after-var": 0, // allow/disallow an empty newline after var statement (off by default)
    "no-array-constructor": 1, // disallow use of the Array constructor
    "no-inline-comments": 1, // disallow comments inline after code (off by default)
    "no-lonely-if": 1, // disallow if as the only statement in an else block (off by default)
    "no-mixed-spaces-and-tabs": 1, // disallow mixed spaces and tabs for indentation
    "no-multiple-empty-lines": [1, {"max": 2}], // disallow multiple empty lines (off by default)
    "no-nested-ternary": 1, // disallow nested ternary expressions (off by default)
    "no-new-object": 1, // disallow use of the Object constructor
    "no-spaced-func": 1, // disallow space between function identifier and application
    "no-ternary": 0, // disallow the use of ternary operators (off by default)
    "no-trailing-spaces": 1, // disallow trailing whitespace at the end of lines
    "no-underscore-dangle": 1, // disallow dangling underscores in identifiers
    "no-wrap-func": 1, // disallow wrapping of non-IIFE statements in parens
    "one-var": [1, "never"], // allow just one var statement per function (off by default)
    "operator-assignment": [1, "never"], // require assignment operator shorthand where possible or prohibit it entirely (off by default)
    "padded-blocks": [1, "never"], // enforce padding within blocks (off by default)
    "quote-props": [1, "as-needed"], // require quotes around object literal property names (off by default)
    "quotes": [1, "single"], // specify whether double or single quotes should be used
    "semi": [1, "always"], // require or disallow use of semicolons instead of ASI
    "semi-spacing": [1, {"before": false, "after": true}], // enforce spacing before and after semicolons
    "sort-vars": 0, // sort variables within the same declaration block (off by default)
    "space-after-keywords": [1, "always"], // require a space after certain keywords (off by default)
    "space-before-blocks": [1, "always"], // require or disallow space before blocks (off by default)
    "space-before-function-paren": [1, {"anonymous": "always", "named": "never"}], // require or disallow space before function opening parenthesis (off by default)
    "space-in-brackets": [1, "never"], // require or disallow spaces inside brackets (off by default)
    "space-in-parens": [1, "never"], // require or disallow spaces inside parentheses (off by default)
    "space-infix-ops": [1, "always"], // require spaces around operators
    "space-return-throw-case": [1, "always"], // require a space after return, throw, and case
    "space-unary-ops": [1, {"words": true, "nonwords": false}], // Require or disallow spaces before/after unary operators (words on by default, nonwords off by default)
    "spaced-line-comment": [1, "always"], // require or disallow a space immediately following the // in a line comment (off by default)
    "wrap-regex": 0, // require regex literals to be wrapped in parentheses (off by default)

    //
    // ECMAScript 6
    //
    // These rules are only relevant to ES6 environments and are off by default.
    //
    "no-var": 2, // require let or const instead of var (off by default)
    "generator-star-spacing": [2, "before"], // enforce the spacing around the * in generator functions (off by default)

    //
    // Legacy
    //
    // The following rules are included for compatibility with JSHint and JSLint.
    // While the names of the rules may not match up with the JSHint/JSLint counterpart,
    // the functionality is the same.
    //
    "max-depth": [2, 3], // specify the maximum depth that blocks can be nested (off by default)
    "max-len": [2, 100, 2], // specify the maximum length of a line in your program (off by default)
    "max-params": [2, 5], // limits the number of parameters that can be used in the function declaration. (off by default)
    "max-statements": 0, // specify the maximum number of statement allowed in a function (off by default)
    "no-bitwise": 0, // disallow use of bitwise operators (off by default)
    "no-plusplus": 2, // disallow use of unary operators, ++ and -- (off by default)

    //
    // eslint-plugin-react
    //
    // React specific linting rules for ESLint
    //
    "react/display-name": 0, // Prevent missing displayName in a React component definition
    "react/jsx-quotes": [2, "double", "avoid-escape"], // Enforce quote style for JSX attributes
    "react/jsx-no-undef": 2, // Disallow undeclared variables in JSX
    "react/jsx-sort-props": 0, // Enforce props alphabetical sorting
    "react/jsx-uses-react": 2, // Prevent React to be incorrectly marked as unused
    "react/jsx-uses-vars": 2, // Prevent variables used in JSX to be incorrectly marked as unused
    "react/no-did-mount-set-state": 2, // Prevent usage of setState in componentDidMount
    "react/no-did-update-set-state": 2, // Prevent usage of setState in componentDidUpdate
    "react/no-multi-comp": 0, // Prevent multiple component definition per file
    "react/no-unknown-property": 2, // Prevent usage of unknown DOM property
    "react/prop-types": 2, // Prevent missing props validation in a React component definition
    "react/react-in-jsx-scope": 2, // Prevent missing React when using JSX
    "react/self-closing-comp": 2, // Prevent extra closing tags for components without children
    "react/wrap-multilines": 2, // Prevent missing parentheses around multilines JSX
  }
}
```

--- a/lib/rules/block-scoped-var.js
+++ b/lib/rules/block-scoped-var.js
@@ -1,337 +1,166 @@
 /**
  * @fileoverview Rule to check for "block scoped" variables by binding context
  * @author Matt DuVall <http://www.mattduvall.com>
+ * @copyright 2015 Toru Nagashima. All rights reserved.
  * @copyright 2015 Mathieu M-Gosselin. All rights reserved.
  */
 "use strict";
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Collects unresolved references from the global scope, then creates a map to references from its name.
+ * @param {RuleContext} context - The current context.
+ * @returns {object} A map object. Its key is the variable names. Its value is the references of each variable.
+ */
+function collectUnresolvedReferences(context) {
+    var unresolved = Object.create(null);
+    var references = context.getScope().through;
+
+    for (var i = 0; i < references.length; ++i) {
+        var reference = references[i];
+        var name = reference.identifier.name;
+
+        if (name in unresolved === false) {
+            unresolved[name] = [];
+        }
+        unresolved[name].push(reference);
+    }
+
+    return unresolved;
+}
+
+/**
+ * Checks whether or not there is a variable of a given name.
+ * @param {escope.Variable[]} variables - An array of variables to be found.
+ * @param {string} name - A name to find.
+ * @returns {boolean} Whether or not there is a variable of the given name.
+ */
+function exists(variables, name) {
+    return variables.some(function(variable) {
+        return variable.name === name;
+    });
+}
 
 //------------------------------------------------------------------------------
 // Rule Definition
 //------------------------------------------------------------------------------
 
 module.exports = function(context) {
-
-    var scopeStack = [];
-
-    //--------------------------------------------------------------------------
-    // Helpers
-    //--------------------------------------------------------------------------
+    var unresolvedReferences = Object.create(null);
+    var stack = [];
 
     /**
-     * Determines whether an identifier is in declaration position or is a non-declaration reference.
-     * @param {ASTNode} id The identifier.
-     * @param {ASTNode} parent The identifier's parent AST node.
-     * @returns {Boolean} true when the identifier is in declaration position.
+     * Makes a block scope.
+     * @param {ASTNode} node - A node of a scope.
+     * @returns {void}
      */
-    function isDeclaration(id, parent) {
-        switch (parent.type) {
-            case "FunctionDeclaration":
-            case "FunctionExpression":
-                return parent.params.indexOf(id) > -1 || id === parent.id;
+    function enterScope(node) {
+        stack.push(node.range);
+    }
 
-            case "VariableDeclarator":
-                return id === parent.id;
+    /**
+     * Pops the last block scope.
+     * @returns {void}
+     */
+    function exitScope() {
+        stack.pop();
+    }
 
-            case "CatchClause":
-                return id === parent.param;
+    /**
+     * Reports a given reference.
+     * @param {escope.Reference} reference - A reference to report.
+     * @returns {void}
+     */
+    function report(reference) {
+        var identifier = reference.identifier;
+        context.report(
+            identifier,
+            "\"{{name}}\" used outside of binding context.",
+            {name: identifier.name});
+    }
 
-            default:
-                return false;
+    /**
+     * Finds and reports references which are outside of valid scopes.
+     * @param {ASTNode} node - A node to get variables.
+     * @returns {void}
+     */
+    function checkForVariables(node) {
+        if (node.kind !== "var") {
+            return;
+        }
+
+        var isGlobal = context.getScope().type === "global";
+
+        // Defines a predicate to check whether or not a given reference is outside of valid scope.
+        var scopeRange = stack[stack.length - 1];
+        function isOutsideOfScope(reference) {
+            var idRange = reference.identifier.range;
+            return idRange[0] < scopeRange[0] || idRange[1] > scopeRange[1];
+        }
+
+        // Gets declared variables, and checks its references.
+        var variables = context.getDeclaredVariables(node);
+        for (var i = 0; i < variables.length; ++i) {
+            var variable = variables[i];
+            var references = variable.references;
+
+            // Global variables are not resolved.
+            // In this case, use unresolved references.
+            if (isGlobal && variable.name in unresolvedReferences) {
+                references = unresolvedReferences[variable.name];
+
+                // Remove processed references.
+                // In order to report references of undefined variables later.
+                delete unresolvedReferences[variable.name];
+            }
+
+            // Reports.
+            references.filter(isOutsideOfScope).forEach(report);
         }
     }
 
     /**
-     * Determines whether an identifier is in property position.
-     * @param {ASTNode} id The identifier.
-     * @param {ASTNode} parent The identifier's parent AST node.
-     * @returns {Boolean} true when the identifier is in property position.
-     */
-    function isProperty(id, parent) {
-        switch (parent.type) {
-            case "MemberExpression":
-                return id === parent.property && !parent.computed;
-
-            case "Property":
-                return id === parent.key;
-
-            default:
-                return false;
-        }
-    }
-
-    /**
-     * Pushes a new scope object on the scope stack.
+     * Finds and reports references which are not resolved (same as no-undef).
      * @returns {void}
      */
-    function pushScope() {
-        scopeStack.push([]);
-    }
+    function checkForUnresolvedReferences() {
+        var globals = context.getScope().variables;
 
-    /**
-     * Removes the topmost scope object from the scope stack.
-     * @returns {void}
-     */
-    function popScope() {
-        scopeStack.pop();
-    }
-
-    /**
-     * Declares the given names in the topmost scope object.
-     * @param {[String]} names A list of names to declare.
-     * @returns {void}
-     */
-    function declare(names) {
-        [].push.apply(scopeStack[scopeStack.length - 1], names);
-    }
-
-    //--------------------------------------------------------------------------
-    // Public API
-    //--------------------------------------------------------------------------
-
-    /**
-     * Declares all relevant identifiers for module imports.
-     * @param {ASTNode} node The AST node representing an import.
-     * @returns {void}
-     * @private
-     */
-    function declareImports(node) {
-        declare([node.local.name]);
-
-        if (node.imported && node.imported.name !== node.local.name) {
-            declare([node.imported.name]);
-        }
-    }
-
-    /**
-     * Declares all relevant identifiers for module exports.
-     * @param {ASTNode} node The AST node representing an export.
-     * @returns {void}
-     * @private
-     */
-    function declareExports(node) {
-        if (node.exported && node.exported.name) {
-            declare([node.exported.name]);
-
-            if (node.local) {
-                declare([node.local.name]);
+        for (var name in unresolvedReferences) {
+            // Skip built-in global variables.
+            // Those references are valid anywhere.
+            if (!exists(globals, name)) {
+                unresolvedReferences[name].forEach(report);
             }
         }
-    }
-
-    /**
-     * Declares all relevant identifiers for classes.
-     * @param {ASTNode} node The AST node representing a class.
-     * @returns {void}
-     * @private
-     */
-    function declareClass(node) {
-
-        if (node.id) {
-            declare([node.id.name]);
-        }
-
-        pushScope();
-    }
-
-    /**
-     * Declares all relevant identifiers for classes.
-     * @param {ASTNode} node The AST node representing a class.
-     * @returns {void}
-     * @private
-     */
-    function declareClassMethod(node) {
-        pushScope();
-
-        declare([node.key.name]);
-    }
-
-    /**
-     * Add declarations based on the type of node being passed.
-     * @param {ASTNode} node The node containing declarations.
-     * @returns {void}
-     * @private
-     */
-    function declareByNodeType(node) {
-
-        var declarations = [];
-
-        switch (node.type) {
-            case "Identifier":
-                declarations.push(node.name);
-                break;
-
-            case "ObjectPattern":
-                node.properties.forEach(function(property) {
-                    declarations.push(property.key.name);
-                    if (property.value) {
-                        declarations.push(property.value.name);
-                    }
-                });
-                break;
-
-            case "ArrayPattern":
-                node.elements.forEach(function(element) {
-                    if (element) {
-                        declarations.push(element.name);
-                    }
-                });
-                break;
-
-            case "AssignmentPattern":
-                declareByNodeType(node.left);
-                break;
-
-            case "RestElement":
-                declareByNodeType(node.argument);
-                break;
-
-            // no default
-        }
-
-        declare(declarations);
-
-    }
-
-    /**
-     * Adds declarations of the function parameters and pushes the scope
-     * @param {ASTNode} node The node containing declarations.
-     * @returns {void}
-     * @private
-     */
-    function functionHandler(node) {
-        pushScope();
-
-        node.params.forEach(function(param) {
-            declareByNodeType(param);
-        });
-
-        declare(node.rest ? [node.rest.name] : []);
-        declare(["arguments"]);
-    }
-
-    /**
-     * Adds declaration of the function name in its parent scope then process the function
-     * @param {ASTNode} node The node containing declarations.
-     * @returns {void}
-     * @private
-     */
-    function functionDeclarationHandler(node) {
-        declare(node.id ? [node.id.name] : []);
-        functionHandler(node);
-    }
-
-    /**
-     * Process function declarations and declares its name in its own scope
-     * @param {ASTNode} node The node containing declarations.
-     * @returns {void}
-     * @private
-     */
-    function functionExpressionHandler(node) {
-        functionHandler(node);
-        declare(node.id ? [node.id.name] : []);
-    }
-
-    function variableDeclarationHandler(node) {
-        node.declarations.forEach(function(declaration) {
-            declareByNodeType(declaration.id);
-        });
-
     }
 
     return {
-        "Program": function() {
-            var scope = context.getScope();
-            scopeStack = [scope.variables.map(function(v) {
-                return v.name;
-            })];
-
-            // global return creates another scope
-            if (context.ecmaFeatures.globalReturn) {
-                scope = scope.childScopes[0];
-                scopeStack.push(scope.variables.map(function(v) {
-                    return v.name;
-                }));
-            }
+        "Program": function(node) {
+            unresolvedReferences = collectUnresolvedReferences(context);
+            stack = [node.range];
         },
 
-        "ImportSpecifier": declareImports,
-        "ImportDefaultSpecifier": declareImports,
-        "ImportNamespaceSpecifier": declareImports,
+        // Manages scopes.
+        "BlockStatement": enterScope,
+        "BlockStatement:exit": exitScope,
+        "ForStatement": enterScope,
+        "ForStatement:exit": exitScope,
+        "ForInStatement": enterScope,
+        "ForInStatement:exit": exitScope,
+        "ForOfStatement": enterScope,
+        "ForOfStatement:exit": exitScope,
+        "SwitchStatement": enterScope,
+        "SwitchStatement:exit": exitScope,
+        "CatchClause": enterScope,
+        "CatchClause:exit": exitScope,
 
-        "ExportSpecifier": declareExports,
-
-        "BlockStatement": function(node) {
-            var statements = node.body;
-            pushScope();
-            statements.forEach(function(stmt) {
-                if (stmt.type === "VariableDeclaration") {
-                    variableDeclarationHandler(stmt);
-                } else if (stmt.type === "FunctionDeclaration") {
-                    declare([stmt.id.name]);
-                }
-            });
-        },
-
-        "VariableDeclaration": function(node) {
-            variableDeclarationHandler(node);
-        },
-
-        "BlockStatement:exit": popScope,
-
-        "CatchClause": function(node) {
-            pushScope();
-            declare([node.param.name]);
-        },
-        "CatchClause:exit": popScope,
-
-        "FunctionDeclaration": functionDeclarationHandler,
-        "FunctionDeclaration:exit": popScope,
-
-        "ClassDeclaration": declareClass,
-        "ClassDeclaration:exit": popScope,
-
-        "ClassExpression": declareClass,
-        "ClassExpression:exit": popScope,
-
-        "MethodDefinition": declareClassMethod,
-        "MethodDefinition:exit": popScope,
-
-        "FunctionExpression": functionExpressionHandler,
-        "FunctionExpression:exit": popScope,
-
-        // Arrow functions cannot have names
-        "ArrowFunctionExpression": functionHandler,
-        "ArrowFunctionExpression:exit": popScope,
-
-        "ForStatement": function() {
-            pushScope();
-        },
-        "ForStatement:exit": popScope,
-
-        "ForInStatement": function() {
-            pushScope();
-        },
-        "ForInStatement:exit": popScope,
-
-        "ForOfStatement": function() {
-            pushScope();
-        },
-        "ForOfStatement:exit": popScope,
-
-        "Identifier": function(node) {
-            var ancestor = context.getAncestors().pop();
-            if (isDeclaration(node, ancestor) || isProperty(node, ancestor) || ancestor.type === "LabeledStatement" || ancestor.type === "BreakStatement" || ancestor.type === "ContinueStatement") {
-                return;
-            }
-
-            for (var i = 0, l = scopeStack.length; i < l; i++) {
-                if (scopeStack[i].indexOf(node.name) > -1) {
-                    return;
-                }
-            }
-
-            context.report(node, "\"" + node.name + "\" used outside of binding context.");
-        }
+        // Finds and reports references which are outside of valid scope.
+        "VariableDeclaration": checkForVariables,
+        "Program:exit": checkForUnresolvedReferences
     };
 
 };

--- a/tests/lib/rules/block-scoped-var.js
+++ b/tests/lib/rules/block-scoped-var.js
@@ -58,7 +58,7 @@ eslintTester.addRuleTest("lib/rules/block-scoped-var", {
         "function f(){ for(var i; i; i) i; }",
         "function f(){ for(var a=0, b=1; a; b) a, b; }",
         "function f(){ for(var a in {}) a; }",
-        "function f(){ switch(2) { case 1: var b = 2; b; break; default: b; break;} b; }",
+        "function f(){ switch(2) { case 1: var b = 2; b; break; default: b; break;} }",
         "a:;",
         "foo: while (true) { bar: for (var i = 0; i < 13; ++i) {if (i === 7) break foo; } }",
         "foo: while (true) { bar: for (var i = 0; i < 13; ++i) {if (i === 7) continue foo; } }",
@@ -75,7 +75,19 @@ eslintTester.addRuleTest("lib/rules/block-scoped-var", {
         { code: "class Test { get flag() { return true; }}", ecmaFeatures: { classes: true }},
         { code: "var Test = class { myFunction() { return true; }}", ecmaFeatures: { classes: true }},
         { code: "var doStuff; let {x: y} = {x: 1}; doStuff(y);", ecmaFeatures: { blockBindings: true, destructuring: true }},
-        { code: "function foo({x: y}) { return y; }", ecmaFeatures: { blockBindings: true, destructuring: true }}
+        { code: "function foo({x: y}) { return y; }", ecmaFeatures: { blockBindings: true, destructuring: true }},
+
+        // https://github.com/eslint/eslint/issues/2253
+        { code: "/*global React*/ let {PropTypes, addons: {PureRenderMixin}} = React; let Test = React.createClass({mixins: [PureRenderMixin]});", ecmaFeatures: { blockBindings: true, destructuring: true }},
+        { code: "/*global prevState*/ const { virtualSize: prevVirtualSize = 0 } = prevState;", ecmaFeatures: { blockBindings: true, destructuring: true }},
+        { code: "const { dummy: { data, isLoading }, auth: { isLoggedIn } } = this.props;", ecmaFeatures: { blockBindings: true, destructuring: true }},
+
+        // https://github.com/eslint/eslint/issues/2747
+        { code: "function a(n) { return n > 0 ? b(n - 1) : \"a\"; } function b(n) { return n > 0 ? a(n - 1) : \"b\"; }"},
+
+        // https://github.com/eslint/eslint/issues/2967
+        { code: "(function () { foo(); })(); function foo() {}"},
+        { code: "(function () { foo(); })(); function foo() {}", ecmaFeatures: { modules: true }}
     ],
     invalid: [
         { code: "!function f(){}; f", errors: [{ message: "\"f\" used outside of binding context." }] },
@@ -106,6 +118,28 @@ eslintTester.addRuleTest("lib/rules/block-scoped-var", {
             code: "function a() { for(var b of {}) { var c = b;} c; }",
             ecmaFeatures: { forOf: true },
             errors: [{ message: "\"c\" used outside of binding context.", type: "Identifier" }]
+        },
+        {
+            code: "function f(){ switch(2) { case 1: var b = 2; b; break; default: b; break;} b; }",
+            errors: [{ message: "\"b\" used outside of binding context.", type: "Identifier" }]
+        },
+        {
+            code: "for (var a = 0;;) {} a;",
+            errors: [{ message: "\"a\" used outside of binding context.", type: "Identifier" }]
+        },
+        {
+            code: "for (var a in []) {} a;",
+            errors: [{ message: "\"a\" used outside of binding context.", type: "Identifier" }]
+        },
+        {
+            code: "for (var a of []) {} a;",
+            ecmaFeatures: { forOf: true },
+            errors: [{ message: "\"a\" used outside of binding context.", type: "Identifier" }]
+        },
+        {
+            code: "{ var a = 0; } a;",
+            ecmaFeatures: { modules: true },
+            errors: [{ message: "\"a\" used outside of binding context.", type: "Identifier" }]
         }
     ]
 });

